### PR TITLE
Replace merge/concat/latest with one join() operator

### LINKED ISSUE #1768: Combine merge/latest/concat into one operator with a “strategy” parameter?
I've been thinking about combining `merge`, `concat`, and `latest` (and their corresponding mapping variants) into one operator—perhaps called `join`—that accepts a choice of joining strategy.

Although it's generally frowned upon to use parameters to _completely_ change the behavior of a function, I think it could be beneficial in this case as a way to force consumers to think about which strategy is most appropriate (vs. RAC 2.x, where almost everyone uses `flattenMap` without thinking about it).

Thoughts?

### PR DESCRIPTION
**Pull request by jspahrsummers**
Resolves #1768.

### CODE REVIEW

**Comment by NachoSoto** [comment]
Looks great!

**Comment by JaviSoto** [comment]
I like it!

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -581,6 +581,80 @@ public func catch<T, E, F>(handler: E -> SignalProducer<T, F>)(producer: SignalP
...
+	///
+	/// The resulting signal will complete only when all inputs have completed.
+	case Concat
+
+	/// Only the events from the latest input signal should be considered for
+	/// the output. Any signals received before that point will be disposed of.
+	///
+	/// The resulting signal will complete only when the signal-of-signals and
+	/// the latest signal has completed.
+	case SwitchToLatest
```
Comment (mdiep) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Could we still call this just _Latest_?

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: :+1: like the succinctness and symmetry to the other values

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 2]
```diff
@@ -581,6 +581,80 @@ public func catch<T, E, F>(handler: E -> SignalProducer<T, F>)(producer: SignalP
 	}
 }
 
+/// Describes how multiple signals or producers should be joined together.
+public enum JoinStrategy: Equatable {
+	/// The signals should be merged, so that any value received on any of the
+	/// input signals will be forwarded immediately to the output signal.
+	///
+	/// The resulting signal will complete only when all inputs have completed.
+	case Merge
```
Comment (joshaber) [inline ReactiveCocoa/Swift/SignalProducer.swift]: In RAC 2, `-flatten:` takes a max number of signals to flatten concurrently (https://github.com/ReactiveCocoa/ReactiveCocoa/blob/54098f8b457d5fa6b2a285d790c2c07c099b98f5/ReactiveCocoa/RACSignal%2BOperations.h#L221). Would we want to parameterize `.Merge` with that?

Comment (mdiep) [inline ReactiveCocoa/Swift/SignalProducer.swift]: I think this would fit better as a separate case. _Limit_, maybe?

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Or `.Concurrent`

Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: I agree that a separate case seems more appropriate.

Also, I wasn't planning on reimplementing that operator for the first Swift release. :flushed:

**Comment by jspahrsummers** [comment]
Ready for review and merge.

**Comment by NachoSoto** [comment]
:shipit:

**Comment by kastiglione** [comment]
I haven't reviewed but a :+1: to a cursor look at the use in the tests :smiley:.

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 3]
```diff
@@ -581,6 +581,80 @@ public func catch<T, E, F>(handler: E -> SignalProducer<T, F>)(producer: SignalP
...
+			return "latest"
+		}
+	}
+}
+
+/// Joins together the inner producers sent upon `producer` according to the
+/// semantics of the given strategy.
+///
+/// If `producer` or any of the inner producers emit an error, the returned
+/// producer will forward that error immediately.
```
Comment (brow) [inline ReactiveCocoa/Swift/SignalProducer.swift]: This isn't quite true. Under the "latest" strategy, an error emitted by an inner producer that is not the latest inner producer is not forwarded.

**Comment by jspahrsummers** [comment]
@brow Thanks for [catching that](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1772#discussion_r25560851)!

Since this got some :+1:s, I'm gonna self-merge it.

**Pull request merged**